turbine/src/cluster_nodes.rs

```diff
@@ -364,13 +363,19 @@ impl<T: 'static> ClusterNodesCache<T> {
     ) -> Arc<ClusterNodes<T>> {
         let epoch = root_bank.get_leader_schedule_epoch(shred_slot);
         let entry = self.get_cache_entry(epoch);
+        if let Some((_, nodes)) = entry
+            .read()
+            .unwrap()
+            .as_ref()
+            .filter(|(asof, _)| asof.elapsed() < self.ttl)
+        {
+            return nodes.clone();
+        }
         // Hold the lock on the entry here so that, if needed, only
         // one thread recomputes cluster-nodes for this epoch.
-        let mut entry = entry.lock().unwrap();
-        if let Some((asof, nodes)) = entry.deref() {
-            if asof.elapsed() < self.ttl {
-                return Arc::clone(nodes);
-            }
+        let mut entry = entry.write().unwrap();
+        if let Some((_, nodes)) = entry.as_ref().filter(|(asof, _)| asof.elapsed() < self.ttl) {
+            return nodes.clone();
```

Comment:
Did you observe much contention on the previous mutex? With the `RwLock`, agreed that this second check of `entry` is necessary to avoid a situation where two threads see a `None` on the earlier check, and then are racing to to populate the entry.

But, wondering if the second check is better than the plain `Mutex`. I guess using the `RwLock` over `Mutex` is optimizing for the cache-hit case (ie allow multiple readers to run in parallel) at the expense of making a cache-miss marginally more expensive.

Reply:
> using the RwLock over Mutex is optimizing for the cache-hit case (ie allow multiple readers to run in parallel) at the expense of making a cache-miss marginally more expensive.

This is precisely the reason for this commit.
99%+ of the time here is cache-hit in which case a write-lock will only add contention.